### ISSUE TITLE
Usage of cryptographically secure RNG for MinGW in NoiseTextures

### ISSUE THREAD

**Issue by m4xw** [gonetz/GLideN64#2236]
Hey @gonetz , rather a old Issue that mudlord fixed on my fork like a 3/4 year ago.
I just started pulling in some updates and saw that enableNoise is gone.
![image](https://user-images.githubusercontent.com/13141469/80697994-1a7b1900-8ada-11ea-80a8-519e0c1f6b7c.png)
This may cause delays up to 20 seconds due to a lack of entropy without our fix under some circumstances.
Probably something you want to deal with upstream?
Reference replaces it with a simple xorshift (https://github.com/libretro/mupen64plus-libretro-nx/blob/develop/GLideN64/src/NoiseTexture.cpp#L38-L44) and there is virtually no visual difference.
I'd be fine doing a PR too.

**Comment by mudlord** [comment]
Problem was:

There was TLS issues with MinGWs implementation of srand and rand(). They use thread local storage which can interfere with how libretro's port of mupen works, thanks to its usage of libco.

Not using anything that relies on threads fixes any stalls. This seems to give enough entropy.

**Comment by loganmc10** [comment]
Yeah I've never noticed any delay in upstream mupen64plus, or heard any complaints, and all the stuff I distribute for Windows uses MinGW

**Comment by gonetz** [comment]
> Reference replaces it with a simple xorshift (https://github.com/libretro/mupen64plus-libretro-nx/blob/develop/GLideN64/src/NoiseTexture.cpp#L38-L44) and there is virtually no visual difference.

![yoda](https://user-images.githubusercontent.com/8983092/80794793-03652580-8bc5-11ea-9e71-89214fcbac44.jpg)

**Comment by gonetz** [comment]
> I'd be fine doing a PR too.

Please do a PR.

**Comment by m4xw** [comment]
@gonetz Is it AL snippet tho?
If anything speaks against it I can implement it from scratch too.
Would you prefer a RNG class that takes in different types or just have it in NoiseTextures like we do for now?

**Comment by mudlord** [comment]
Yes it is.

**Comment by mudlord** [comment]
> Yeah I've never noticed any delay in upstream mupen64plus, or heard any complaints, and all the stuff I distribute for Windows uses MinGW

Prob because it doesn't rely on cothreads or how libco does cothreads.

**Comment by gonetz** [comment]
> If anything speaks against it I can implement it from scratch too.

No, I don't think that it is a problem.

**Comment by m4xw** [comment]
@gonetz `Would you prefer a RNG class that takes in different types or just have it in NoiseTextures like we do for now?`
If you don't answer that I will just slap the commit 1:1 in a PR ;)

**Comment by gonetz** [comment]
I don't know what is RNG class. `inline u32 irand()` in NoiseTextures is fine for me.

**Comment by m4xw** [comment]
```c
RNG* generator = new RNG(RNG::CryptographicallySecure)
u32 number = generator->randomU32();
delete generator;
generator = new RNG(RNG::XorShift);
number = generator->randomU32();
delete generator;
```
Could look something like this and we can refactor it in GLideN64.
Could also do it as a singleton.

**Comment by gonetz** [comment]
RNG class is ok. It should be lightweight, so singleton is not necessary.

**Comment by gonetz** [comment]
@m4xw Please do a PR.

**Comment by m4xw** [comment]
@gonetz I plan do after I got my mupen next release wrapped up, its currently stalling since a month because we are investigating a nasty issue in new_dynarec and I also currently work on a Citra port to Switch.
Sorry I didn't forget it, I just have my hands full :/

**Comment by gonetz** [comment]
> Sorry I didn't forget it, I just have my hands full :/

It's ok, nothing urgent here.

**Comment by m4xw** [comment]
Considering resolved for now as per #2757
